src/platform/CYW30739/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,213 @@
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    if (!wiced_firmware_upgrade_prepare())
+    {
+        ChipLogError(SoftwareUpdate, "wiced_firmware_upgrade_prepare");
+        return;
+    }
+
+    imageProcessor->mParams.downloadedBytes = 0;
+
+    imageProcessor->mDownloader->OnPreparedForDownload(CHIP_NO_ERROR);
+}
+
+void OTAImageProcessorImpl::HandleFinalize(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    if (!wiced_firmware_upgrade_finalize())
+    {
+        ChipLogError(SoftwareUpdate, "Failed to finalize OTA upgrade");
+        return;
+    }
+
+    imageProcessor->ReleaseBlock();
+
+    ChipLogProgress(SoftwareUpdate, "OTA image downloaded");
+}
+
+void OTAImageProcessorImpl::HandleApply(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    if (!wiced_firmware_upgrade_apply())
+    {
+        ChipLogError(SoftwareUpdate, "Failed to apply OTA upgrade");
+        return;
+    }
+
+    ChipLogProgress(SoftwareUpdate, "OTA upgrade completed");
+
+    wiced_hal_wdog_reset_system();
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    wiced_firmware_upgrade_abort();
+
+    ChipLogError(SoftwareUpdate, "OTA upgrade was aborted");
+
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    const uint32_t written = wiced_firmware_upgrade_process_block(imageProcessor->mParams.downloadedBytes,
+                                                                  imageProcessor->mBlock.data(), imageProcessor->mBlock.size());
+    if (written != imageProcessor->mBlock.size())
+    {
+        ChipLogError(SoftwareUpdate, "wiced_firmware_upgrade_process_block 0x%08lx", written);
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+        return;
+    }
+
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
+{
+    if ((block.data() == nullptr) || block.empty())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    // Allocate memory for block data if it has not been done yet
+    if (mBlock.empty())
```

Comment:
This implementation of SetBlock is buggy.  Please copy one of the non-buggy ones (e.g. Linux).

Reply:
Fixed.